File: solutions/dynamic-programming/062-unique-paths.py
I'll help you implement a solution for the Unique Paths problem in Python. This solution will use dynamic programming to calculate the number of unique paths from the top-left corner to the bottom-right corner of a grid.

```python
#!/usr/bin/env python3
"""
Unique Paths - Dynamic Programming Solution
-----------------------------------------
Problem: Calculate the number of unique paths from top-left to bottom-right corner
of an m x n grid, where you can only move right or down at any point.

Time Complexity: O(m*n)
Space Complexity: O(m*n) - can be optimized to O(n)
"""

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        """
        Calculate the number of unique paths in an m x n grid.
        
        Args:
            m (int): Number of rows in the grid
            n (int): Number of columns in the grid
            
        Returns:
            int: Number of unique paths from top-left to bottom-right
        """
        # Create a 2D DP table initialized with 1s
        dp = [[1] * n for _ in range(m)]
        
        # Fill the DP table
        # For each cell, paths = paths from above + paths from left
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i-1][j] + dp[i][j-1]
        
        # Return the bottom-right cell value
        return dp[m-1][n-1]
    
    def uniquePathsOptimized(self, m: int, n: int) -> int:
        """
        Space-optimized version using 1D array instead of 2D array.
        
        Args:
            m (int): Number of rows in the grid
            n (int): Number of columns in the grid
            
        Returns:
            int: Number of unique paths from top-left to bottom-right
        """
        # Create a 1D array for current row
        dp = [1] * n
        
        # Process each row
        for i in range(1, m):
            # Update each cell in the row
            for j in range(1, n):
                dp[j] += dp[j-1]
        
        return dp[n-1]


def test_unique_paths():
    """
    Test function to verify the implementation.
    """
    solution = Solution()
    
    # Test cases
    test_cases = [
        (3, 7, 28),
        (3, 2, 3),
        (7, 3, 28),
        (3, 3, 6),
        (1, 1, 1),
    ]
    
    for m, n, expected in test_cases:
        result = solution.uniquePaths(m, n)
        optimized_result = solution.uniquePathsOptimized(m, n)
        assert result == expected, f"Failed for m={m}, n={n}. Expected {expected}, got {result}"
        assert optimized_result == expected, f"Optimized version failed for m={m}, n={n}"
        print(f"Passed: m={m}, n={n}, paths={result}")


if __name__ == "__main__":
    # Run tests
    test_unique_paths()
    
    # Example usage
    solution = Solution()
    m, n = 3, 7
    result = solution.uniquePaths(m, n)
    print(f"\nNumber of unique paths in a {m}x{n} grid: {result}")
```

This implementation includes:

1. A standard solution using 2D dynamic programming (`uniquePaths`)
2. An optimized solution using 1D array (`uniquePathsOptimized`)
3. Comprehensive test cases
4. Clear documentation and comments
5. Proper Python conventions and structure

The solution uses dynamic programming to build up the number of possible paths to each cell. The key insight is that the number of paths to any cell is the sum of paths from the cell above it and the cell to its left.

The optimized version reduces space complexity from O(m*n) to O(n) by using only a single row array and updating it in place.

To use this code:
1. Save it to the specified path
2. Run it directly to execute the tests
3. Import the Solution class to use in other code

The code handles edge cases and includes assertions to verify correct functionality.

File: solutions/dynamic-programming/062-unique-paths.js
I'll help you create a complete JavaScript implementation for the Unique Paths problem. I'll write the solution with clear comments and proper structure.

```javascript
/**
 * Unique Paths Problem Solution
 * 
 * Problem: A robot is located at the top-left corner of a m x n grid.
 * The robot can only move either down or right at any point in time.
 * The robot is trying to reach the bottom-right corner of the grid.
 * How many possible unique paths are there?
 * 
 * @param {number} m - Number of rows in the grid
 * @param {number} n - Number of columns in the grid
 * @return {number} - Number of unique paths
 */

/**
 * Dynamic Programming solution for the Unique Paths problem
 * Time Complexity: O(m*n)
 * Space Complexity: O(m*n)
 */
function uniquePaths(m, n) {
    // Handle edge cases
    if (m <= 0 || n <= 0) return 0;
    if (m === 1 || n === 1) return 1;

    // Create a 2D array to store the number of paths for each cell
    const dp = Array(m).fill().map(() => Array(n).fill(0));

    // Initialize first row with 1 as there's only one way to reach each cell in first row
    for (let j = 0; j < n; j++) {
        dp[0][j] = 1;
    }

    // Initialize first column with 1 as there's only one way to reach each cell in first column
    for (let i = 0; i < m; i++) {
        dp[i][0] = 1;
    }

    // Fill the dp array
    // For each cell, paths = paths from above + paths from left
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            dp[i][j] = dp[i-1][j] + dp[i][j-1];
        }
    }

    // Return the value at bottom-right cell
    return dp[m-1][n-1];
}

/**
 * Space-optimized solution using 1D array
 * Time Complexity: O(m*n)
 * Space Complexity: O(n)
 */
function uniquePathsOptimized(m, n) {
    // Handle edge cases
    if (m <= 0 || n <= 0) return 0;
    if (m === 1 || n === 1) return 1;

    // Create 1D array for current row
    const dp = Array(n).fill(1);

    // Process each row
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            dp[j] += dp[j-1];
        }
    }

    return dp[n-1];
}

// Test cases
function runTests() {
    const testCases = [
        { m: 3, n: 7, expected: 28 },
        { m: 3, n: 2, expected: 3 },
        { m: 7, n: 3, expected: 28 },
        { m: 1, n: 1, expected: 1 },
        { m: 2, n: 2, expected: 2 }
    ];

    console.log("Running tests...");
    testCases.forEach((test, index) => {
        const result = uniquePaths(test.m, test.n);
        const optimizedResult = uniquePathsOptimized(test.m, test.n);
        console.log(`Test ${index + 1}:`);
        console.log(`Input: m = ${test.m}, n = ${test.n}`);
        console.log(`Expected: ${test.expected}`);
        console.log(`Regular Solution Result: ${result}`);
        console.log(`Optimized Solution Result: ${optimizedResult}`);
        console.log(`Status: ${result === test.expected ? 'PASSED' : 'FAILED'}`);
        console.log('---');
    });
}

// Run the tests
runTests();

// Export the functions for external use
module.exports = {
    uniquePaths,
    uniquePathsOptimized
};
```

This implementation includes:

1. Two solutions for the Unique Paths problem:
   - A standard dynamic programming solution using 2D array
   - A space-optimized solution using 1D array

2. Detailed comments explaining the approach and complexity

3. Proper error handling for edge cases

4. A comprehensive test suite to verify the implementation

5. Module exports for external use

The code follows JavaScript best practices and conventions, including:
- Consistent naming conventions
- Clear function and variable names
- Proper spacing and formatting
- Comprehensive documentation
- Error handling
- Test cases

The solution uses dynamic programming to solve the problem efficiently, with both a standard implementation and a space-optimized version. The space-optimized version reduces the space complexity from O(m*n) to O(n) while maintaining the same time complexity.